File: scripts/DataProcessor.py
from pathlib import Path
import os

class DataProcessor:
    """
    A tool to batch process CSV files by converting content to uppercase.

    This class provides a structured way to handle directory-level file 
    transformations, ensuring that the output folder is prepared and
    files are handled using modern Pathlib logic.

    Attributes:
        input_dir (Path): Source directory containing the raw .csv files.
        output_dir (Path): Target directory where processed results are saved.
    """

    def __init__(self, input_dir, output_dir):
        """
        Initializes the DataProcessor with input and output paths.

        Args:
            input_dir (str or Path): The directory containing files to process.
            output_dir (str or Path): The directory where output files will be written.
        """
        self.input_dir = Path(input_dir)
        self.output_dir = Path(output_dir)
        
        # Action: Create the folder if it doesn't exist
        self.output_dir.mkdir(parents=True, exist_ok=True)

    def process_file(self, file_path):
        """
        Reads a single file and converts its text content to uppercase.

        Args:
            file_path (Path): The Path object representing the file to be read.

        Returns:
            str: The entire content of the file converted to uppercase strings.
        """
        # .read_text() is a modern 'pathlib' way to open and read in one go
        data = file_path.read_text()
        return data.upper()

    def run(self):
        """
        Orchestrates the batch processing of all CSV files in the input directory.

        This method identifies all files ending in .csv, applies the transformation 
        logic via process_file(), and writes the new content to the output directory 
        with a '_processed' suffix.

        Returns:
            None
        """
        # .glob("*.csv") finds all matching files in the folder
        for path in self.input_dir.glob("*.csv"):
            result = self.process_file(path)
            
            # Use / operator (pathlib magic) to join paths safely
            # .stem gets the filename without the .csv extension
            out_path = self.output_dir / f"{path.stem}_processed.csv"
            
            out_path.write_text(result)
            print(f"Done: {path.name} -> {out_path.name}")
            
            
#### Running the processor 

data_dir = os.path.join("..", "data")
out_dir = os.path.join("..", "outputs")

if __name__ == "__main__":
    # Example usage
    processor = DataProcessor(input_dir="data", output_dir="outputs")
    processor.run()

File: scripts/DataProcessor.R
library(R6)
library(fs) # The 'pathlib' of R

#' @title DataProcessor Class
#' @description A class to batch process CSV files by converting content to uppercase.
#'
#' @field input_dir Path. The directory containing raw CSV files.
#' @field output_dir Path. The directory where processed files will be saved.
#' @export
DataProcessor <- R6Class("DataProcessor",
  public = list(
  input_dir = NULL,
  output_dir = NULL,

#' @description 
#' The Constructor: Initializes paths and creates the output folder.
#' @param input_dir Character or Path. Directory containing raw data.
#' @param output_dir Character or Path. Directory for processed results.
  initialize = function(input_dir, output_dir) {
    # Use fs::path() for cross-platform safety (like Path())
    self$input_dir <- path(input_dir)
    self$output_dir <- path(output_dir)
    
    # Action: Create the folder if it doesn't exist
    dir_create(self$output_dir)
  },

  #' @description
  #' Logic for a single file: reads a CSV and converts content to uppercase.
  #' @param file_path Character or Path. The specific file to process.
  #' @return A data frame with all character columns converted to uppercase.
  process_file = function(file_path) {
    data <- read.csv(file_path)
    # Note: toupper() on a whole dataframe converts it to a character vector.
    # For a workshop, this logic is usually applied to specific columns.
    return(toupper(data)) 
  },

  #' @description
  #' The main 'Engine': Orchestrates the batch processing across the input directory.
  #' @return None. Writes files to the output directory and prints messages.
  run = function() {
    # dir_ls with a glob pattern finds the files
    files <- dir_ls(self$input_dir, glob = "*.csv")
    
    for (f in files) {
      result <- self$process_file(f)
      
      # path_file/path_ext_remove are like .name and .stem in pathlib
      file_stem <- path_ext_remove(path_file(f))
      out_path <- path(self$output_dir, paste0(file_stem, "_processed.csv"))
      
      write.csv(result, out_path, row.names = FALSE)
      message(paste("Done:", path_file(f), "->", path_file(out_path)))
    }
  })
)


### --- How to use it --- ###

### The input directories
data_dir <- path("data")
out_dir  <- path("outputs")

processor <- DataProcessor$new(input_dir = data_dir, output_dir = out_dir)
processor$run()

## Diagnostic tools for the processor class
str(processor)
ls(processor)

# Viewing the input variables that have been selected.
processor$input_dir
